### ISSUE TITLE
weird results from reverse( %x = reverse %h )

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7526]
Migrated from [rt.perl.org#31865](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=31865) (status was 'resolved')

Searchable as RT31865$

**Comment by p5pRT** [comment]
### From bzm@2bz.de

#### Created by bzm@2bz.de



lisa&#8203;:\~ borisz$ perl \-MData&#8203;::Dumper \-le '%h = \(1 => 2\, a=>"b"\, c => 2\);  
%h = reverse \( %x = reverse %h\); print Dumper\(\{x => \\%x\, h =>\\%h\}\);'
$VAR1 = \{
&nbsp;           'h' => \{
&nbsp;                    '' => 2
&nbsp;                  \}\,
&nbsp;           'x' => \{
&nbsp;                    'b' => 'a'\,
&nbsp;                    '2' => 'c'
&nbsp;                  \}
&nbsp;         \};


Note that %h is damaged now\. I really expect to loose the dupes\.

Some more opinions here&#8203;:
http&#8203;://www\.perlmonks\.com/index\.pl?node\_id=394658

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=medium

Site configuration information for perl v5.8.4:

Configured by sherm at Sat Apr 24 12:35:38 EDT 2004.

Summary of my perl5 (revision 5 version 8 subversion 4) configuration:
   Platform:
     osname=darwin, osvers=7.3.0, archname=darwin-thread-multi-2level
     uname='darwin sherm-pendleys-computer.local 7.3.0 darwin kernel  
version 7.3.0: fri mar 5 14:22:55 pst 2004;  
root:xnuxnu-517.3.15.obj~4release_ppc power macintosh powerpc '
     config_args='-Dprefix=/usr/local/cb/perl5.8.4 -Dusethreads  
-Duseshrplib -Dcf_email=perl@dot-app.org -des'
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=define use5005threads=undef useithreads=define  
usemultiplicity=define
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=undef use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-pipe -fno-common -DPERL_DARWIN -no-cpp-precomp  
-fno-strict-aliasing -I/usr/local/include',
     optimize='-Os',
     cppflags='-no-cpp-precomp -pipe -fno-common -DPERL_DARWIN  
-no-cpp-precomp -fno-strict-aliasing -I/usr/local/include'
     ccversion='', gccversion='3.3 20030304 (Apple Computer, Inc. build  
1495)', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',  
lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags ='  
-L/usr/local/lib'
     libpth=/usr/local/lib /usr/lib
     libs=-ldbm -ldl -lm -lc
     perllibs=-ldl -lm -lc
     libc=/usr/lib/libc.dylib, so=dylib, useshrplib=true,  
libperl=libperl.dylib
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_dyld.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup  
-L/usr/local/lib'

Locally applied patches:



@INC for perl v5.8.4:
     /sw/lib/perl5
     /sw/lib/perl5/darwin
     /usr/local/cb/perl5.8.4/lib/5.8.4/darwin-thread-multi-2level
     /usr/local/cb/perl5.8.4/lib/5.8.4
      
/usr/local/cb/perl5.8.4/lib/site_perl/5.8.4/darwin-thread-multi-2level
     /usr/local/cb/perl5.8.4/lib/site_perl/5.8.4
     /usr/local/cb/perl5.8.4/lib/site_perl
     .


Environment for perl v5.8.4:
     DYLD_LIBRARY_PATH (unset)
     HOME=/Users/borisz
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
      
PATH=/sw/bin:/sw/sbin:/usr/local/cb/perl5.8.4/bin:/usr/local/bin:/bin:/ 
sbin:/usr/bin:/usr/sbin:/usr/X11R6/bin
     PERL5LIB=/sw/lib/perl5:/sw/lib/perl5/darwin
     PERL_BADLANG (unset)
     SHELL=/bin/bash
--
Boris


```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Boris Zentner \(via RT\) wrote&#8203;:
> lisa&#8203;:\~ borisz$ perl \-MData&#8203;::Dumper \-le '%h = \(1 => 2\, a=>"b"\, c => 2\);  
> %h = reverse \( %x = reverse %h\); print Dumper\(\{x => \\%x\, h =>\\%h\}\);'
> $VAR1 = \{
>            'h' => \{
>                     '' => 2
>                   \}\,
>            'x' => \{
>                     'b' => 'a'\,
>                     '2' => 'c'
>                   \}
>          \};
> 
> 
> Note that %h is damaged now\. I really expect to loose the dupes\.

Another "stuff on the stack is not refcounted" bug ?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From rick@bort.ca

On Wed\, Oct 06\, 2004 at 06&#8203;:50&#8203;:17PM \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> Boris Zentner \(via RT\) wrote&#8203;:
> > lisa&#8203;:\~ borisz$ perl \-MData&#8203;::Dumper \-le '%h = \(1 => 2\, a=>"b"\, c => 2\);  
> > %h = reverse \( %x = reverse %h\); print Dumper\(\{x => \\%x\, h =>\\%h\}\);'
> > $VAR1 = \{
> >            'h' => \{
> >                     '' => 2
> >                   \}\,
> >            'x' => \{
> >                     'b' => 'a'\,
> >                     '2' => 'c'
> >                   \}
> >          \};
> > 
> > 
> > Note that %h is damaged now\. I really expect to loose the dupes\.
> 
> Another "stuff on the stack is not refcounted" bug ?

I dunno\, but some more data points&#8203;:

&nbsp;    $ perl5\.8\.3 \-le 'print \(%x = \("a"\,1\,"b"\,2\,"b"\,3\)\);'
&nbsp;    bb
&nbsp;    $ perl5\.6\.1 \-le 'print \(%x = \("a"\,1\,"b"\,2\,"b"\,3\)\);'
&nbsp;    a1bb3

I'm not sure what it should print but those are both obviously wrong\.

\-\- 
Rick Delaney
rick@&#8203;bort\.ca

**Comment by p5pRT** [comment]
### From @gbarr


On 6 Oct 2004\, at 17&#8203;:50\, Rafael Garcia\-Suarez wrote&#8203;:

> Boris Zentner \(via RT\) wrote&#8203;:
>> lisa&#8203;:\~ borisz$ perl \-MData&#8203;::Dumper \-le '%h = \(1 => 2\, a=>"b"\, c => 2\);
>> %h = reverse \( %x = reverse %h\); print Dumper\(\{x => \\%x\, h =>\\%h\}\);'
>> $VAR1 = \{
>>            'h' => \{
>>                     '' => 2
>>                   \}\,
>>            'x' => \{
>>                     'b' => 'a'\,
>>                     '2' => 'c'
>>                   \}
>>          \};
>>
>>
>> Note that %h is damaged now\. I really expect to loose the dupes\.
>
> Another "stuff on the stack is not refcounted" bug ?

I am not so sure\. I do not have a blead perl to run right now\, but 
compare

perl \-MDevel&#8203;::Peek \-e '%h = \(1 => 2\, a=>"b"\, c => 2\); Dump\($\_\) for @&#8203;x = 
reverse %h'

perl \-MDevel&#8203;::Peek \-e '%h = \(1 => 2\, a=>"b"\, c => 2\); Dump\($\_\) for %x = 
reverse %h'

It seems that what is left on the stack after a hash assign in a list 
context is bad\.

But reading the code it seems to me that the code above would not do 
what the user expected anyway\. The code in pp\_aassign looks like it is 
trying to do the same thing regardless of the LHS being an array or a 
hash\. So it would seem the stack would have the same contents as the 
RHS and not whats left after the hash assignment had removed 
duplicates\.

Graham\.

**Comment by p5pRT** [comment]
### From @demerphq


> I dunno\, but some more data points&#8203;:
> 
>     $ perl5\.8\.3 \-le 'print \(%x = \("a"\,1\,"b"\,2\,"b"\,3\)\);'
>     bb
>     $ perl5\.6\.1 \-le 'print \(%x = \("a"\,1\,"b"\,2\,"b"\,3\)\);'
>     a1bb3
> 
> I'm not sure what it should print but those are both obviously wrong\.

Maybe its also OS specific? On Win32 I couldn't repeat this behaviour using
either AS Perl 5\.6\.1 \(658\) or under a home built Perl 5\.8\.5 nor under the
Cygwin Perl 5\.8\.5 either\.

D&#8203;:\\Development>perl \-le "print \(%x = \(qw\(a 1 b 2 c 3\)\)\);"
a1b2c3

D&#8203;:\\Development>perl \-v

This is perl\, v5\.6\.1 built for MSWin32\-x86\-multi\-thread
\(with 1 registered patch\, see perl \-V for more detail\)

Copyright 1987\-2001\, Larry Wall

Binary build 638 provided by ActiveState Corp\. http&#8203;://www\.ActiveState\.com
ActiveState is a division of Sophos\.
Built Apr 13 2004 19&#8203;:24&#8203;:21

D&#8203;:\\Development>perl5\.8\.5 \-le "print \(%x = \(qw\(a 1 b 2 c 3\)\)\);"
a1b2c3

D&#8203;:\\Development>e&#8203;:\\perl5\.8\.5\\bin\\perl \-le "print \(%x = \(qw\(a 1 b 2 c 3\)\)\);"
a1b2c3

D&#8203;:\\Development>perl5\.8\.5 \-v

This is perl\, v5\.8\.5 built for cygwin\-thread\-multi\-64int

Copyright 1987\-2004\, Larry Wall

Perl may be copied only under the terms of either the Artistic License or
the
GNU General Public License\, which may be found in the Perl 5 source kit\.

Complete documentation for Perl\, including FAQ lists\, should be found on
this system using \`man perl' or \`perldoc perl'\.  If you have access to the
Internet\, point your browser at http&#8203;://www\.perl\.com/\, the Perl Home Page\.


D&#8203;:\\Development\\rating\\csb\_sp>e&#8203;:\\perl5\.8\.5\\bin\\perl \-v

This is perl\, v5\.8\.5 built for MSWin32\-x86\-multi\-thread
\(with 1 registered patch\, see perl \-V for more detail\)

Copyright 1987\-2004\, Larry Wall

Perl may be copied only under the terms of either the Artistic License or
the
GNU General Public License\, which may be found in the Perl 5 source kit\.

Complete documentation for Perl\, including FAQ lists\, should be found on
this system using \`man perl' or \`perldoc perl'\.  If you have access to the
Internet\, point your browser at http&#8203;://www\.perl\.com/\, the Perl Home Page\.

**Comment by p5pRT** [comment]
### From @demerphq

> Maybe its also OS specific? On Win32 I couldn't repeat this behaviour
using
> either AS Perl 5\.6\.1 \(658\) or under a home built Perl 5\.8\.5 nor under the
> Cygwin Perl 5\.8\.5 either\.

Er\, I meant AS 638\. Sorry\.

Also it looks like AS 638 gets it more right but not correct than later
versions&#8203;:

D&#8203;:\\Development>perl5\.6\.1 \-MData&#8203;::Dumper \-le "%h = \(1 => 2\, a=>'b'\, c => 2\);
%h = reverse \( %x = reverse %h\); pr
int Dumper\(\{x => \\%x\, h =>\\%h\}\);"
$VAR1 = \{
&nbsp;          'h' => \{
&nbsp;                   '' => 2\,
&nbsp;                   '1' => 2\,
&nbsp;                   'a' => 'b'
&nbsp;                 \}\,
&nbsp;          'x' => \{
&nbsp;                   '2' => '1'\,
&nbsp;                   'b' => 'a'
&nbsp;                 \}
&nbsp;        \};


D&#8203;:\\Development>perl5\.8\.5 \-MData&#8203;::Dumper \-le "%h = \(1 => 2\, a=>'b'\, c => 2\);
%h = reverse \( %x = reverse %h
\); print Dumper\(\{x => \\%x\, h =>\\%h\}\);"
$VAR1 = \{
&nbsp;          'h' => \{
&nbsp;                   '' => 2
&nbsp;                 \}\,
&nbsp;          'x' => \{
&nbsp;                   'b' => 'a'\,
&nbsp;                   '2' => 'c'
&nbsp;                 \}
&nbsp;        \};

**Comment by p5pRT** [comment]
### From rick@bort.ca

On Wed\, Oct 06\, 2004 at 06&#8203;:58&#8203;:58PM \+0100\, Graham Barr wrote&#8203;:
> 
> But reading the code it seems to me that the code above would not do 
> what the user expected anyway\. The code in pp\_aassign looks like it is 
> trying to do the same thing regardless of the LHS being an array or a 
> hash\. So it would seem the stack would have the same contents as the 
> RHS and not whats left after the hash assignment had removed 
> duplicates\.

It looks like it's trying to treat arrays and hashes differently to me\,
in particular note the comment&#8203;:

&nbsp;        /\* Removes from the stack the entries which ended up as
&nbsp;         \* duplicated keys in the hash \(fix for \[perl \#24380\]\) \*/

Maybe that fix just didn't have enough test cases\.  As for what it
should do\, I just found in perlop

&nbsp;       Similarly\, a list assignment in list context produces the
&nbsp;       list of lvalues assigned to\, and a list assignment in
&nbsp;       scalar context returns the number of elements produced by
&nbsp;       the expression on the right hand side of the assignment\.

So I think Boris's code should do what he expected\.

\-\- 
Rick Delaney
rick@&#8203;bort\.ca

**Comment by p5pRT** [comment]
### From rick@bort.ca

On Wed\, Oct 06\, 2004 at 07&#8203;:00&#8203;:59PM \+0100\, Orton\, Yves wrote&#8203;:
> 
> > I dunno\, but some more data points&#8203;:
> > 
> >     $ perl5\.8\.3 \-le 'print \(%x = \("a"\,1\,"b"\,2\,"b"\,3\)\);'
> >     bb
> >     $ perl5\.6\.1 \-le 'print \(%x = \("a"\,1\,"b"\,2\,"b"\,3\)\);'
> >     a1bb3
> > 
> > I'm not sure what it should print but those are both obviously wrong\.
> 
> Maybe its also OS specific? On Win32 I couldn't repeat this behaviour using
> either AS Perl 5\.6\.1 \(658\) or under a home built Perl 5\.8\.5 nor under the
> Cygwin Perl 5\.8\.5 either\.
> 
> D&#8203;:\\Development>perl \-le "print \(%x = \(qw\(a 1 b 2 c 3\)\)\);"
&nbsp;                                                   ^
&nbsp;                                                   b in my example
> a1b2c3

The required trigger is a duplicate key which you don't have in your
trials\.

\-\- 
Rick Delaney
rick@&#8203;bort\.ca

**Comment by p5pRT** [comment]
### From @demerphq

> > D&#8203;:\\Development>perl \-le "print \(%x = \(qw\(a 1 b 2 c 3\)\)\);"
>                                                    ^
>                                                    b in my example
> > a1b2c3
> 
> The required trigger is a duplicate key which you don't have in your
> trials\.

Whoops\. So I didn't sorry for the noise \(I got the same results as you once
the key was duped\)\.

Yves

**Comment by p5pRT** [comment]
### From @gbarr


On 6 Oct 2004\, at 19&#8203;:11\, Rick Delaney wrote&#8203;:

> On Wed\, Oct 06\, 2004 at 06&#8203;:58&#8203;:58PM \+0100\, Graham Barr wrote&#8203;:
>>
>> But reading the code it seems to me that the code above would not do
>> what the user expected anyway\. The code in pp\_aassign looks like it is
>> trying to do the same thing regardless of the LHS being an array or a
>> hash\. So it would seem the stack would have the same contents as the
>> RHS and not whats left after the hash assignment had removed
>> duplicates\.
>
> It looks like it's trying to treat arrays and hashes differently to me\,
> in particular note the comment&#8203;:
>
>         /\* Removes from the stack the entries which ended up as
>          \* duplicated keys in the hash \(fix for \[perl \#24380\]\) \*/

Yes\, you are right\. I was looking at old source\.

But the code seems flawed\. It counts duplicates during the assignment 
loop\.
Then at the end does a single move on the stack\. The move should happen
during assignment loop\.

Graham\.

**Comment by p5pRT** [comment]
### From @davidnicol

here's the arguably obvious workaround\, since the bug applies to the
list returned
from assignment of a list to a hash\, don't use
assignment\-of\-an\-array\-to\-a\-hash as an
rvalue\.  One assumes that Boris did this to get the result he wanted\.

$ perl \-MData&#8203;::Dumper \-le '%h = \(1 => 2\, a=>"b"\, c => 2\); %x = reverse %h; %h =
 reverse %x; print Dumper\(\{x => \\%x\, h =>\\%h\}\);'
$VAR1 = \{
&nbsp;          'h' => \{
&nbsp;                   'c' => '2'\,
&nbsp;                   'a' => 'b'
&nbsp;                 \}\,
&nbsp;          'x' => \{
&nbsp;                   'b' => 'a'\,
&nbsp;                   '2' => 'c'
&nbsp;                 \}
&nbsp;        \};

> 
> lisa&#8203;:\~ borisz$ perl \-MData&#8203;::Dumper \-le '%h = \(1 => 2\, a=>"b"\, c => 2\);
> %h = reverse \( %x = reverse %h\); print Dumper\(\{x => \\%x\, h =>\\%h\}\);'
> $VAR1 = \{
>            'h' => \{
>                     '' => 2
>                   \}\,
>            'x' => \{
>                     'b' => 'a'\,
>                     '2' => 'c'
>                   \}
>          \};
> 
> Note that %h is damaged now\. I really expect to loose the dupes\.


\-\- 
David L Nicol
"Valuable ideas can withstand scrutiny" \-\- authors of /XP Explained/

**Comment by p5pRT** [comment]
### From @iabyn

Fixed in blead with commit 4596056478d3ae4ae183d2821eb95156aff83924

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
